File: Data Structures & Algorithms/merge-intervals/submission-6.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda x: x[0])
        res = [intervals[0]]

        for start, end in intervals:
        # such a simple problem just have to traverse the list and update the laste end by checking if there is an overlap otherwise just
        # add the new segment into the arrayu
            last_end = res[-1][1]
            if start <= last_end:
                res[-1][1] = max(last_end, end)
            else:
                res.append([start, end])
        return res

File: Data Structures & Algorithms/valid-parenthesis-string/submission-6.cpp
class Solution {
public:
    bool checkValidString(string s) {
        int low = 0, high = 0;
        for (char a: s) {
            if (a == '(') {
                low++;
                high++;
            } else if (a == ')') {
                low--;
                high--;
            } else {
                low--;
                high++;
            }
            if (high < 0) { return false; }
            low = max(low, 0);
        }
        return low == 0;
    }
};
